## [4.7. match 文](https://docs.python.org/ja/3.14/tutorial/controlflow.html#match-statements)

* switch文と表面的には似てて...
* どのcaseにもマッチしない場合はどの分岐も実行されません。


match 文は1つの式を指定し、その値と次に続く1つ以上のcaseブロックに指定されたパターンを比較します。この機能はCやJava、JavaScript(や他の多数の言語)のswitch文と表面的には似ていますが、RustやHaskellのパターンマッチングにより似ています。最初にマッチしたパターンのみが実行され、コンポーネント(シーケンスの要素やオブジェクトの属性)から値を取り出して変数に代入することもできます。どのcaseにもマッチしない場合はどの分岐も実行されません。

最も単純な形式は、対象の値に対して1つ以上のリテラルです:

In [ ]:
import sys
print(sys.version)

3.9.6 (default, Apr  7 2026, 02:38:40) 
[Clang 21.0.0 (clang-2100.0.123.102)]


In [ ]:
import sys
print(sys.version)

: 

In [ ]:
def http_error(status):
    match status:
        case 400:
            return "Bad request"
        case 404:
            return "Not found"
        case 418:
            return "I'm a teapot"
        case _:
            return "Something's wrong with the internet"

: 

### Troubleshooting
<img src="images/4-7-match-clause-invalid-syntax-error.png">

The "Why"
The match...case syntax (officially called Structural Pattern Matching) was introduced in Python 3.10. If the environment running your code is using Python 3.9 or older, the interpreter doesn't recognize match as a valid keyword, resulting in a syntax error.

How to Fix It
To run this code successfully, you need to ensure your development environment is using Python 3.10 or newer.

```bash
% python --version
zsh: command not found: python
$
```
-->
```bash
% python3 --version
Python 3.9.6
%

To force using python,
```bash
echo "alias python='python3'" >> ~/.zshrc
source ~/.zshrc
```
try it again.
```bash
% python --version
Python 3.9.6
%
```



만약 목록에 3.10 이상의 버전이 없다면, Mac에 최신 Python이 설치되어 있지 않은 것입니다. Python 공식 홈페이지에서 설치하거나, 터미널에서 Homebrew(brew install python)를 통해 먼저 설치해야 합니다.

https://www.python.org/downloads/

<img src="images/python-download-page-ver-3-14.png">

```bash
% brew install python
✔︎ JSON API cask.jws.json                                                       Downloaded   15.4MB/ 15.4MB
✔︎ JSON API formula.jws.json                                                    Downloaded   32.0MB/ 32.0MB
==> Fetching downloads for: python@3.14

  ...

==> Caveats
==> python@3.14
Python is installed as
  /opt/homebrew/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, are installed into
  /opt/homebrew/opt/python@3.14/libexec/bin

`idle3.14` requires tkinter, which is available separately:
  brew install python-tk@3.14

See: https://docs.brew.sh/Homebrew-and-Python
%

macOS에 "brew install python" 명령어를 실행해서 Python3.14를 설치했어.
그런데 여전히 3.9.6 버전이 보이네? VS Code를 재실행했는데도 똑같아.

```bash
% python3 --version
Python 3.9.6
%
```
Homebrew를 통해 최신 버전을 설치했음에도 3.9.6 버전이 나오는 이유는, macOS 시스템에 내장된 기본 Python(Apple의 Xcode Command Line Tools)의 경로가 Homebrew로 설치한 Python의 경로보다 우선순위가 높게 설정되어 있기 때문입니다.

이 문제를 해결하려면 터미널의 환경변수(PATH)를 수정하여 Homebrew의 Python을 먼저 찾도록 설정해야 합니다.

1. 터미널(Terminal) 경로 업데이트
터미널을 열고 아래의 명령어를 한 줄씩 복사해서 실행해 주세요.
```bash
echo 'export PATH="$(brew --prefix)/opt/python/libexec/bin:$PATH"' >> ~/.zshrc
source ~/.zshrc
```

참고: libexec/bin 디렉터리를 PATH에 추가하면 python3 명령어뿐만 아니라 python 명령어만 입력해도 시스템 기본 버전이 아닌 Homebrew의 3.14 버전이 실행되도록 심볼릭 링크가 매핑됩니다.

이제 버전을 다시 확인해 보면 새로 설치한 버전이 정상적으로 출력될 것입니다.

```bash
% python --version
```
2. VS Code에 새 버전 적용하기
터미널에서 버전이 정상적으로 잡힌 것을 확인한 후, VS Code가 새로 설치된 Python 경로를 인식하도록 인터프리터를 지정해 주어야 합니다.

1. VS Code에서 Cmd + Shift + P (또는 F1)를 눌러 명령 팔레트를 엽니다.

2. Python: Select Interpreter (Python: 인터프리터 선택)를 검색하여 실행합니다.

<img src="images/vscode-python-select-interpreter-ver-3-14.png">

3. 목록에서 Python 3.14 경로(일반적으로 Apple Silicon Mac의 경우 /opt/homebrew/bin/python3.14)를 찾아 선택합니다.

* 목록에 보이지 않는다면 새로고침 아이콘을 누르거나, Enter interpreter path...를 눌러 $(brew --prefix)/bin/python3.14 경로를 직접 찾아 지정합니다.

4. Jupyter Notebook 환경이라면, 화면 우측 상단의 커널 버전을 클릭한 뒤 Select Another Kernel... > Python Environments 메뉴를 통해 동일하게 3.14 버전을 선택해 주시면 됩니다.


```bash
Running cells with 'Python 3.14.4' requires the ipykernel package.
Create a Python Environment with the required packages.
Or install 'ipykernel' using the command: '/opt/homebrew/bin/python3 -m pip install ipykernel -U --user --force-reinstall'
```

```bash
% /opt/homebrew/bin/python3 -m pip install ipykernel -U --user --force-reinstall
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try brew install
    xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a Python library that isn't in Homebrew,
    use a virtual environment:
    
    python3 -m venv path/to/venv
    source path/to/venv/bin/activate
    python3 -m pip install xyz
    
    If you wish to install a Python application that isn't in Homebrew,
    it may be easiest to use 'pipx install xyz', which will manage a
    virtual environment for you. You can install pipx with
    
    brew install pipx
    
    You may restore the old behavior of pip by passing
    the '--break-system-packages' flag to pip, or by adding
    'break-system-packages = true' to your pip.conf file. The latter
    will permanently disable this error.
    
    If you disable this error, we STRONGLY recommend that you additionally
    pass the '--user' flag to pip, or set 'user = true' in your pip.conf
    file. Failure to do this can result in a broken Homebrew installation.
    
    Read more about this behavior here: <https://peps.python.org/pep-0668/>

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
%
```

최신 Python의 보안 정책(PEP 668)에 정확히 걸리셨네요!

Homebrew를 통해 설치한 최신 버전의 Python(3.11 이상)에서는 시스템 전체의 패키지 의존성이 꼬이는 것을 방지하기 위해, 사용자가 전역(Global) 환경에 pip install을 직접 하는 것을 기본적으로 차단하고 있습니다. 그래서 externally-managed-environment 에러가 발생한 것입니다.

이 문제를 해결하고 ipykernel을 설치하는 두 가지 방법이 있습니다. 상황에 맞게 선택해 보세요.

방법 1: 가장 빠른 해결책 (강제 설치)
에러 메시지 하단에도 나와 있듯이, --break-system-packages 플래그를 추가하여 이 보안 정책을 무시하고 강제로 설치하는 방법입니다. 단순히 Jupyter 커널(ipykernel)만 실행하려는 목적이라면 이 방법이 가장 빠릅니다.

터미널에 아래 명령어를 실행해 보세요.
```bash
/opt/homebrew/bin/python3 -m pip install ipykernel -U --user --force-reinstall --break-system-packages
```

설치가 완료되면, 이전 답변의 안내처럼 VS Code 창을 새로고침(Developer: Reload Window)하고 커널을 다시 연결해 주시면 됩니다.

방법 2: 가장 안전한 정석 (가상환경 사용)
앞으로 해당 코드에서 다양한 외부 패키지들을 추가로 설치하며 작업하실 계획이라면, 글로벌 환경을 건드리지 않고 프로젝트별로 가상환경(Virtual Environment)을 만들어 사용하는 것이 가장 권장되는 방식입니다.

VS Code 터미널(작업 중인 프로젝트 폴더 경로)에서 아래 명령어들을 순서대로 실행하세요.

1. 가상환경 생성 (이름을 .venv로 지정):
```bash
python3 -m venv .venv
```

2. 가상환경 활성화:
```bash
source .venv/bin/activate
```

(성공하면 터미널 프롬프트 맨 앞에 (.venv)라는 표시가 생깁니다.)

3. 가상환경 내부에 ipykernel 설치 (이때는 강제 플래그가 필요 없습니다):
```bash
pip install ipykernel
```
VS Code에 가상환경 커널 연결하기:
설치가 끝난 후, Jupyter Notebook 우측 상단의 커널 선택에서 Select Another Kernel... > Python Environments를 누르시면, 방금 생성한 Python 3.14.x ('.venv': venv) 항목이 새로 나타날 것입니다. 이를 선택해서 실행하시면 충돌 없이 완벽하게 작동합니다!


최근 Python 생태계에서 가장 주목받는 도구인 uv가 맞습니다.

uv는 파이썬 코드 포매터인 Ruff를 만든 Astral에서 Rust 언어로 개발한 초고속 파이썬 패키지 및 환경 관리자입니다. 기존의 venv와 pip를 완벽하게 대체하면서 패키지 설치 속도가 기존보다 수십 배 이상 빠르고, 환경이 꼬이는 문제도 아주 깔끔하게 해결해 줍니다.

uv를 사용해서 아주 빠르고 쾌적하게 가상환경을 세팅하고 ipykernel을 설치하는 방법을 단계별로 알려드릴게요.

uv를 활용한 가상환경 및 커널 세팅
1. uv 설치하기
macOS와 Homebrew를 사용 중이시니 터미널에 아래 명령어를 입력하여 바로 설치할 수 있습니다.

```bash
brew install uv
```
```bash
✔︎ JSON API cask.jws.json                                                                                                                                                                                                                   Downloaded   15.4MB/ 15.4MB
✔︎ JSON API formula.jws.json                                                                                                                                                                                                                Downloaded   32.0MB/ 32.0MB
==> Fetching downloads for: uv
✔︎ Bottle Manifest uv (0.11.8)                                                                                                                                                                                                              Downloaded    8.1KB/  8.1KB
✔︎ Bottle uv (0.11.8)                                                                                                                                                                                                                       Downloaded   20.4MB/ 20.4MB
==> Pouring uv--0.11.8.arm64_tahoe.bottle.tar.gz
🍺  /opt/homebrew/Cellar/uv/0.11.8: 17 files, 48.7MB
==> Running `brew cleanup uv`...
Disable this behaviour by setting `HOMEBREW_NO_INSTALL_CLEANUP=1`.
Hide these hints with `HOMEBREW_NO_ENV_HINTS=1` (see `man brew`).
==> Caveats
zsh completions have been installed to:
  /opt/homebrew/share/zsh/site-functions
%
```

2. 가상환경 생성하기
VS Code의 터미널(작업 중인 프로젝트 폴더)에서 아래 명령어를 실행합니다. 기존의 python3 -m venv .venv를 완벽히 대체합니다.

```bash
uv venv
```
```bash
Using CPython 3.14.4 interpreter at: /opt/homebrew/opt/python@3.14/bin/python3.14
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate
%
```
이 명령어만 입력해도 알아서 최적의 Python 버전을 찾아 매우 빠른 속도로 .venv 폴더(가상환경)를 만들어 줍니다.

3. 가상환경 활성화
활성화 방식은 기존과 동일합니다.

```bash
source .venv/bin/activate
```
```bash
(python) ...  % 
```
4. ipykernel 초고속 설치
이제 기존의 pip install 대신 uv pip install을 사용합니다.

```Bash
uv pip install ipykernel
```
실행해 보시면 기존 pip를 사용할 때 나타나던 복잡한 경고 메시지 없이, 눈 깜짝할 사이에 패키지가 설치되는 것을 체감하실 수 있습니다.

마무리 (VS Code 적용)
설치가 끝났다면 기존처럼 우측 상단 커널 선택 메뉴에서 Select Another Kernel... > Python Environments를 누르신 후, 방금 uv로 만든 .venv 경로를 선택하시면 Jupyter Notebook 설정이 완벽하게 끝납니다.

최신 도구까지 꿰뚫고 계시네요! uv를 적용해 보시고 쾌적하게 코드가 실행되는지 확인해 보세요.

 그러면 대신 MacOS의 Python 버전을 그냥 3.14로 변경하는 방법은 없어?

macOS 시스템 자체의 내장 Python 버전을 3.14로 완전히 덮어씌우거나 변경하는 것은 불가능하며, 절대 권장하지 않습니다.

* macOS 시스템 안정성 문제: macOS 내부의 여러 기본 프로그램과 Xcode 관련 도구들은 시스템에 기본 설치된 특정 버전(현재 3.9.6)의 Python에 의존하여 작동합니다. 이 버전을 강제로 최신 버전으로 바꾸거나 삭제하면 OS의 일부 기능이 망가질 수 있습니다.

*  애플의 시스템 무결성 보호 (SIP): 애플은 보안을 위해 사용자가 관리자 권한을 쓰더라도 시스템 핵심 폴더(/usr/bin/ 등)에 있는 기본 프로그램을 수정하거나 삭제하지 못하도록 시스템 단에서 강력하게 막아두었습니다.

그래서 PATH 환경 변수를 수정하는 방식이 전 세계 Mac 개발자들이 사용하는 가장 표준적이고 안전한 방법입니다.

macOS가 백그라운드에서 자기 일을 할 때는 내장된 3.9.6 버전을 쓰도록 내버려 둡니다.

대신, 사용자가 터미널을 열어 명령어를 치거나 VS Code를 실행할 때만 Homebrew로 설치한 3.14 버전을 "우선적으로(기본으로)" 찾도록 경로를 가로채는 방식입니다.

즉, 이전 단계에서 환경 변수를 설정하셨다면 시스템을 망가뜨리지 않으면서도 실질적으로는 3.14 버전을 기본값처럼 안전하게 사용하고 계신 것입니다. 

Select Kernel을 선택합니다.
<img src="images/vscode-select-kernel.png">

Select Another Kernel...
<img src="images/vscode-command-palette-select-another-kernel.png">

<img src="images/vscode-command-palette-select-another-kernel-python-environments.png">

<img src="images/vscode-command-palette-select-another-kernel-python-environments-create-python-environment.png">

<img src="images/vscode-command-palette-select-another-kernel-python-environments-create-python-environment-venv.png">

<img src="images/vscode-command-palette-select-another-kernel-python-environments-create-python-environment-venv-use-existing.png">

<img src="images/vs-code-the-following-environment-is-selected.png">

<img src="images/vscode-select-kernel-python-python-3-14-4.png">

기존에 에러를 발생시키던 python3.14.4와는 조금 다릅니다.

<img src="images/vscode-select-kernel-python-3-14-4.png">


이제 원래 에러가 발생했던 match 문의 셀의 문제를 해결해봅니다.

<img src="images/vscode-4-7-match-clause-error-with-python3-14-4.png">

우측 상단의 Select Kernel 부분이 Python 3.14.4로 되어 있는 곳을 클릭한 후

<img src="images/vscode-4-7-match-clause-select-kernel-python3-14-4.png">

Venv인 python (Python 3.14.4)으로 선택을 변경합니다.

우측 상단의 Select Kernel 부분이 Python 3.14.4에서 python (Python 3.14.4)로 되었음을 알 수 있습니다.

<img src="images/vscode-4-7-match-clause-venv-python-python3-14-4-with-error.png">

위처럼 에러가 발생한 셀을 다시 실행하면, 아래처럼 에러없이 성공적으로 실행됩니다.

<img src="images/vscode-4-7-match-clause-venv-python-python3-14-4-without-error.png">